### DISCUSSION TITLE
How to view custom Generated Vector Tiles

### DISCUSSION THREAD

**Discussion by onmax** [maplibre/maplibre-gl-js#3594] [Q&A]
Good morning everyone. First of all, thanks for all of the contributors of Map Libre, really great project! 

# TLDR

**Do you know any tool that I can use to see my Custom Vector Tiles?** I want to check if they are being created correctly :)

# The problem

- ✅ I have a map is being rendered
- ✅ The Nuxt endpoint seems to be returning values.
- ❌ I cannot see my Vector Tiles in the map

# My setup

I have Postgres database (Supabase) and `PostGIS` enabled. I have a table called `gas_stations` and there is one column `geolocation` that contain the `lat` and `lng` information of the gas stations.

My goal is to render all the gas stations in the map. Since I have over 11,000 locations, I was thinking to use Vector Tiles. 

## Function in Postgres (Supabase)

I wrote a Postgres function to generate tiles and return them given the `zxy` params. It seems that it is returning values.

<details><summary><code>get_gas_station_tile</code> function</summary>
<p>

```psql
DROP function get_gas_station_tile;
CREATE OR REPLACE FUNCTION public.get_gas_station_tile(z integer, x integer, y integer)
RETURNS bytea AS $$
BEGIN
    RETURN (
        SELECT ST_AsMVT(q, 'gas_stations', 4096, 'mvt_geom')
        FROM (
            SELECT 
                id,
                name,
                address,
                diesel,
                gasoline_95_e5,
                opening_hours,
                ST_AsMVTGeom(
                    geolocation::geometry,
                    TileBBox(z, x, y, 4326),
                    4096,
                    256,
                    false
                ) AS mvt_geom
            FROM 
                public.gas_stations
            WHERE 
                geolocation && TileBBox(z, x, y, 4326)
        ) AS q
    );
END;
$$ LANGUAGE plpgsql STABLE;
```

</p>
</details> 

## Nuxt API endpoint

From my `nuxt` `server/api`, I have [created a function](https://pastebin.com/Fa8PLN5Q) that receives the `zxy` parameters, make the request to the postgres database, convert the returned value to binary and return it. This also seems to be working.

<details><summary>Function in my Nuxt server <code>/tiles/{z}/{x}/{y}</code></summary>
<p>

```ts
// /api/tiles/[...zxy].ts
import { Buffer } from 'node:buffer'
import { serverSupabaseClient } from '#supabase/server'
import type { Database } from '~/types/supabase'
 
export default defineEventHandler(async (event) => {
  const params = event.context.params?.zxy
  if (!params || params.split('/').length !== 3)
    return new Response('Invalid params', { status: 400 })
 
  const [z, x, y] = params.split('/').map(Number)
 
  if (Number.isNaN(z) || Number.isNaN(x) || Number.isNaN(y))
    return new Response('Invalid params', { status: 400 })
 
  const client = await serverSupabaseClient<Database>(event)
  const { data, error } = await client.rpc('get_gas_station_tile', { z, x, y })
 
  if (error)
    return new Response(JSON.stringify(error), { status: 500 })
  const binaryData = Buffer.from(data, 'hex')
  return new Response(binaryData, { status: 200, headers: { 'Content-Type': 'application/x-protobuf' } })
})
```

</p>
</details> 

## Rendering the Vector Tile using MapLibre

<details><summary>Code in the client</summary>
<p>

```js
const map = new maplibregl.Map({
    container: 'map',
    style: 'https://api.maptiler.com/maps/streets/style.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL',
    zoom: 6,
    center: [-3, 40.447303],
  })
  map.on('load', () => {
    map.addSource('gas_stations', {
      type: 'vector',
      tiles: ['http://localhost:3000/api/tiles/{z}/{x}/{y}'],
    })
    map.addLayer({
      'id': 'gas_stations',
      'type': 'circle',
      'source': 'gas_stations',
      'source-layer': 'gas_stations', // Make sure this matches the layer name in your tiles
      'paint': {
        'circle-radius': 6, // Adjust the size of the circle
        'circle-color': '#ff0000', // Red color for the circle
      },
    })
  })
```
</p>
</details> 

This seems to be working as I am not getting any error/warning in the console.

# My question

I am bit lost on how to debug this issue. I don't even know if my tiles are being created correctly, so I would like to start from there, but not sure how can I check that. I tried to use https://observablehq.com/@kotelnikov/vector-tiles-viewer, but I get this error: `Unimplemented type: 6`

1. **Do you know any tool that I can use to see my tiles?**
2. Do you spot something in my code that is not correct or it can be improved?

Thank you so much!

**Comment by onmax** [comment]
I am trying https://maputnik.github.io/, but it is not loading.

I tried adding two sources: From my Supabase and from my localhost.

But I have the same issue, the points are not being redendered.

**Reply by acalcutt** [reply]
You would not be able to use localhost from an external url/page ... you need a public url for that to work.

**Reply by onmax** [reply]
Yes, I am using `--tunel` in the nuxt cli, which uses [untun](https://github.com/unjs/untun)

I test the request with https://hoppscotch.io/ and it is returning things

**Comment by acalcutt** [comment]
One tool I have used to troubleshoot tiles is [vt2geojson](https://github.com/mapbox/vt2geojson) . It helps you see a decoded geojson version of what is in your mvt tile, which you can load with other tools to see what you have and troubleshoot.

For example, in this issue I used to to look at the polygons generated in google earth
https://github.com/maplibre/maplibre-native/issues/1843#issuecomment-1799388230

**Reply by onmax** [reply]
Cool. I will try it. Thanks

**Comment by acalcutt** [comment]
I think "Unimplemented type: 6" may mean your tiles are gzipped, but your not setting the gzipped headers. For example, in tileserver-gl the 'gzip' 'Content-Encoding' header gets added to gzipped content ( https://github.com/maptiler/tileserver-gl/blob/master/src/serve_data.js#L95 )

**Reply by acalcutt** [reply]
Or looking at the code again, maybe it just needs to be gzipped + those headers set?

**Reply by onmax** [reply]
I will play with that :)

**Reply by acannistra** [reply]
This is the answer I believe.

**Reply by onmax** [reply]
I don't think so :/

I tried this in my Nuxt API

```ts
const { data, error } = await client.rpc('get_gas_station_tile', { z, x, y })

if (error)
  return new Response(JSON.stringify(error), { status: 500 })
const binaryData = Buffer.from(data, 'hex')
  
const headers: HeadersInit = {
  'Content-Type': 'application/x-protobuf',
  'Content-Encoding': 'gzip',
  'Content-Length': `${binaryData.length}`,
  'Cache-Control': 'public, max-age=86400',
}
return new Response(binaryData, { status: 200, headers })
```

**Reply by acannistra** [reply]
Does `binaryData` represent a gzipped vector tile?

**Reply by onmax** [reply]
I updated the code.

For some reason I think the response from supabase is being converted to hexadecimal, so I need to put it in binary again. Maybe the error is there.

**Reply by acannistra** [reply]
If `binaryData` is a gzipped representation of a vector tile, the `Content-Encoding: gzip` value is correct. If it is not, do not provide `Content-Encoding: gzip` in the set of headers.

**Reply by onmax** [reply]
And do you know if it is required to set any header when serving pbf files?

**Reply by acannistra** [reply]
I can't quite remember, to be honest. I am confident that `Content-Type` is relevant, but I honestly can't quite remember if `application/x-protobuf`, `application/octet-stream` or `application/vnd.mapbox-vector-tile` is the right value. 

`Content-Encoding` is definitely important to be set correctly, because the browser handles decompression. If it's set incorrectly (or not set when it should be), the data provided from the response to the maplibre renderer will be wrong.

**Reply by acalcutt** [reply]
+1 on the `Content-Encoding` being important. I know when I hosted tiles on github pages, I had to uncompress them so they would work there, since `Content-Encoding` isn't set when tiles are served from there (like step 3 from https://github.com/maplibre/demotiles/tree/gh-pages/tiles-omt#omt-tile-generation which uncompressed the tiles so they could be served that way)

Not sure if 'Content-Type' is important, but know tileserver uses application/x-protobuf 
https://github.com/maptiler/tileserver-gl/blob/master/src/serve_data.js#L130

**Reply by acalcutt** [reply]
If you look at the 'network' tab in developer tools in chrome, you might be able to verify those header are set, like

![image](https://github.com/maplibre/maplibre-gl-js/assets/3792408/678a20d4-da99-4f92-b24b-4440bc182f99)
